### DISCUSSION TITLE
Use WASM shared memory?

### DISCUSSION THREAD

**Discussion by infogulch** [ncruces/go-sqlite3#69] [Ideas]
As of two months ago, wazero has symbols that mention [shared memory](https://github.com/tetratelabs/wazero/blame/253c0346e4c8d7bd604304447233c1a3ce0a71f7/internal/wazeroir/compiler.go#L225). Could this be used to lift some of the caveats mentioned in the [readme](https://github.com/ncruces/go-sqlite3?tab=readme-ov-file#write-ahead-logging) such as allowing locking mode `NORMAL` and connection pooling?

**Comment by infogulch** [comment]
I'm quite satisfied to get issue number 69, nice! 😄

Sorry, did you intend to mark this discussion "Closed as resolved"?

**Reply by ncruces** [reply]
No, I'm sorry.

I intended to start a discussion of the possibilities, instead of an issue that tracks progress… or lack thereof.

I've had WAL as a goal from the beginning, started and deleted a couple of design docs, jumped at every new avenue, including this one (there's a [threads](https://github.com/ncruces/go-sqlite3/tree/threads) branch).

I also looked into shared cache as a mitigation.

I honestly don't think it's feasible with the current state of affairs, but I'm open to suggestions.

I want to keep this relatively simple to maintain, and in particular, I want to remain able update to new SQLite versions in less than a day's work.

**Reply by infogulch** [reply]
No problem, I was just a little confused by the misclick. Perhaps I could have been more patient. 😄

I can appreciate your goal of designing this library to be simple enough to keep up to date easily.

**Comment by ncruces** [comment]
That's a great question, with a long answer.

The simplest answer to your question is: yes, _kinda_. It could certainly help; but with its own set of caveats.

I explicitly mention WAL in the [README](https://github.com/ncruces/go-sqlite3?tab=readme-ov-file#write-ahead-logging) because it's considered best practice to use it, and as such, I'd really like to support it, to as full an extent as possible.

As the README states, I created a from scratch VFS which has a bunch of advantages and compromises.

So, in the current architecture, each SQLite connection lives in its own little sandbox, completely isolated from other connections. Through the VFS layer, Go acts like an OS kernel, managing these OS “processes.” There are no threads, no locks, no shared cache, etc. It's a good model.

Unfortunately, SQLite expects to be able to share memory between these “processes” for the WAL index (just a few pages) which we can't do.

With WASM shared memory another model is possible. We create a big (4GB) shared memory, and we run all our SQLite connections on it. But, if we want to have any concurrency, we need threads, locks, etc.

At the time of this writing, this is possible to compile with wasi-sdk, and to run with wazero; [go-re2](https://github.com/wasilibs/go-re2) uses this model. 

But… it's complicated. The ABI to setup threads is far from obvious or stable. wazero APIs, are likewise, experimental. Moving to this model would for sure involve API redesign for this package. And support for platforms other than 64-bit Linux/Windows/macOS is dubious at best.

Finally, this still won't allow our SQLite connections to share memory with SQLite connections from other processes, like Litestream. It's also an open question how to lock the database to ensure other processes couldn't access it while we have multiple connections open (it's a change to the locking protocol that needs a design).

I've also considered other options. I could create a VFS where each connection thinks it's in memory journal mode, but then I send transactions to the WAL. Or I could look into the [libSQL](https://github.com/tursodatabase/libsql) virtual WAL.

All solutions are more work than I want to take in at the moment, and ultimately, maintain.

**Reply by infogulch** [reply]
I see there are a lot of strategies in this space with different tradeoffs.

> Unfortunately, SQLite expects to be able to share memory between these “processes” for the WAL index (just a few pages) which we can't do.

Why is this not possible? This is what I would expect wazero's shared memory would enable: just enough memory shared between different WASM module instances in the same Go process to satisfy SQLite's shared memory expectation.

**Reply by ncruces** [reply]
A WASM module uses a linear memory, which it declares or imports.
In the more common scenario, each module creates its own memory, isolated from other modules'.

But you can also have one module that creates a memory, and the other modules import that memory: they'll all share the same memory. But it's still one memory: one address space they'll have to share.

For something like SQLite, **complied from C**, this address space will store all global variables, all heap memory, all thread stacks. You can't really have different global variables, so all your modules need to be the same WASM. `malloc` and `free` need to coordinate (synchronize) between modules. Thread stacks need to be `malloc`ed from the heap, and setup appropriately. All according to the C compiler ABI.

All this is finicky, especially because the wasi-threads folks have said they [don't endorse it](https://github.com/WebAssembly/wasi-threads/issues/45).

You _seem_ to think we could have one module having multiple memories, some shared, some private?
That's [another proposal](https://github.com/WebAssembly/multi-memory), that I don't think wazero supports, and [it's unclear](https://github.com/WebAssembly/wasi-sdk/issues/211) how it could be supported by the C compiler.

**Reply by infogulch** [reply]
> You seem to think we could have one module having multiple memories, some shared, some private?

Ah I see. Yes you correctly intuited my misunderstanding. I didn't realize that the shared memory feature is all or nothing; maybe I could have guessed based on the diff.

Doesn't sqlite support multiprocess shared cache? I'm not sure how this is set up by sqlite, but assuming its something like mmap I'd imagine a multiple memory WASM environment overrides whatever that is use a shared memory instead.

**Reply by ncruces** [reply]
No, shared cache is single process, and [discouraged](https://sqlite.org/sharedcache.html#use_of_shared_cache_is_discouraged).

**Reply by infogulch** [reply]
Ah. It appears that SQLite expects the VFS to implement shared memory primitives:

> ### [Write-Ahead Logging](https://sqlite.org/wal.html)
>
> * WAL normally requires that the [VFS](https://sqlite.org/vfs.html) support shared-memory primitives.
> ...
> * There is an additional quasi-persistent "-wal" file and "-shm" shared memory file associated with each database, which can make SQLite less appealing for use as an [application file-format](https://sqlite.org/appfileformat.html).
> ...
> #### [Implementation Of Shared-Memory For The WAL-Index](https://sqlite.org/wal.html#implementation_of_shared_memory_for_the_wal_index)
>
> Specialized applications for which the default implementation of shared memory is unacceptable can devise alternative methods via a custom [VFS](https://sqlite.org/vfs.html). For example, if it is known that a particular database will only be accessed by threads within a single process, the wal-index can be implemented using heap memory instead of true shared memory.

> #### WAL-mode File Format - [The Wal-Index or "-shm" file](https://sqlite.org/walformat.html#the_wal_index_or_shm_file)
> 
> Because the shm is only used to coordinate access between concurrent clients, the shm file is omitted if [exclusive locking mode](https://sqlite.org/pragma.html#pragma_locking_mode) is set, as an optimization. When [exclusive locking mode](https://sqlite.org/pragma.html#pragma_locking_mode) is set, SQLite uses heap memory in place of the memory-mapped shm file.

Using mmap in Go is not unheard of:

- [Exploring MMAP In Go](https://ghvsted.com/blog/exploring-mmap-in-go/)
- [Russ Cox' use of mmap](https://github.com/google/codesearch/blob/master/index/mmap_linux.go#L13)

Could we use mmap to map a *-shm file multiple times, once for each instance?

**Reply by ncruces** [reply]
No, we couldn't: we can't mmap into the guest's memory.

To mmap into the guests memory, I'd need to mmap into a specific address of the guests memory slice. Few platforms allow mmaping into a chosen address, and those that do, take it only as a hint. Additionally, from that point forward, the memory slice could not be safely appended to any longer (guest memory can't grow).

This is not a simple problem to solve.

**Reply by infogulch** [reply]
I'm sorry if you find my line of questioning to be annoying. I'm just trying to explore the solution space, not assert that an easy solution exists. Thank you for your patience so far. If you're tired of responding, well, fair enough 🤷😅.

I'm still fuzzy on what SQLite's VFS requires for shared memory, but I would expect that it doesn't require mapping to a specific address. I.e. you tell it what address the shared memory can be found, it doesn't tell you where it wants it.

It seems like the "strategy" for using mmap would be to first have the kernel find some large contiguous address space with `MAP_ANONYMOUS | MAP_PRIVATE` that is the size of the entire guest memory, and then call mmap again within that region with `MAP_FIXED | MAP_SHARED` to map a portion of the file (at a system page size offset) which is much more likely to succeed due to the previous mmap call. This is the strategy used by [wuffs](https://github.com/google/wuffs/blob/main/script/mmap-ring-buffer.c#L87) and the article [Using black magic to make a fast circular buffer.](https://lo.calho.st/posts/black-magic-buffer/) and the linux man page for [mmap(2)](https://man7.org/linux/man-pages/man2/mmap.2.html) under the heading *Using MAP_FIXED safely* ([?](https://stackoverflow.com/a/76860466/213246)).

And well what do you know, wazero is already using mmaped memory for common platforms: [internal/wasm/memory.go:NewMemoryInstance](https://github.com/tetratelabs/wazero/blob/main/internal/wasm/memory.go#L67-L109); [mmap_unix.go](https://github.com/tetratelabs/wazero/blob/main/internal/platform/mmap_unix.go#L17); [mmap_windows.go](https://github.com/tetratelabs/wazero/blob/main/internal/platform/mmap_windows.go#L61); [mmap_unsupported.go](https://github.com/tetratelabs/wazero/blob/main/internal/platform/mmap_unsupported.go#L26).

Maybe wazero could expose an api to try to map a file to a region within the guest's mmaped memory and maintain it, or produce an error if it can't.

**Reply by ncruces** [reply]
> I'm sorry if you find my line of questioning to be annoying.

Questions are fine, I want to explore too. It's just stuff like _“mmap in Go is not unheard of”_ and _“what do you know, wazero uses mmap”_ that makes it harder than it needs to be. I mean, I know `mmap` can be used from Go, and I'm a wazero (community, not core) maintainer.

I also know `MAP_FIXED` and the caveats to use it safely. You do, however make a good point (that's great, thanks!): if wazero `mmap`s, it's safer for me to `MAP_FIXED`. wazero didn't `mmap` when I discarded the idea.

It's still very unsafe to do, however. wazero only `mmap`s under certain conditions. Namely, memory size needs to be fixed, not elastic. I'd also need to ensure the bit I remap can't be used by something else, like `malloc`. I'd need to `malloc` something huge, then `mmap` it? And I have no idea how to implement unmap to release the file.

Also no Windows and maybe no macOS.

Those are just the first few caveats to pop into my head.

**Reply by ncruces** [reply]
[go-re2](https://github.com/wasilibs/go-re2) were probably the primary drivers of wazero `mmap`ing memory. Just two days after [v1.5.0](https://github.com/wasilibs/go-re2/releases/tag/v1.5.0) was released, they had to release [v1.5.1](https://github.com/wasilibs/go-re2/releases/tag/v1.5.1) to fix allocating too much _virtual_ memory, because memory is now fixed size. The [fix](https://github.com/wasilibs/go-re2/pull/79) works, but is exactly the kind of thing I'd rather avoid.

I don't mind exploring a amd64/arm64 Linux only solution. It [won't](https://github.com/ncruces/go-sqlite3/blob/d6db364271726e35830773e2e1b3fe7724b5f6c0/vfs/os_f2fs_linux.go) be the first.

But there's a couple of things I want to ensure/avoid. I want to make sure the lib works on memory limited containers. I want to make sure it still works with the [interpreter,](https://github.com/ncruces/go-sqlite3/blob/d6db364271726e35830773e2e1b3fe7724b5f6c0/.github/workflows/test.yml#L113) and on [weird](https://github.com/ncruces/go-sqlite3/blob/d6db364271726e35830773e2e1b3fe7724b5f6c0/.github/workflows/cross.sh) platforms. And I'd much rather not to have to create a different WASM binary specifically to target WAL.

On this last point I'm asking if it's possible to trigger wazero `mmap` optionally, without requiring shared memory. It's still annoying, since it potentially needs a second [runtime](https://github.com/ncruces/go-sqlite3/blob/d6db364271726e35830773e2e1b3fe7724b5f6c0/sqlite.go#L32) which I'd also rather avoid.

**Reply by ncruces** [reply]
I'll use this wazero [PR](https://github.com/tetratelabs/wazero/pull/2157) to experiment.

If I can work around the memory management issues - implement `xShmMap` _and_ `xShmUnmap` sensibly - `xShmLock` and `xShmBarrier` should be simple.

I don't think this will ship in the _“default”_ configuration, however. It's full of caveats.

**Reply by ncruces** [reply]
`syscall.Mmap` doesn't work because it creates the slice. Need to use `go:linkname`, if I want this to work on macOS. Tracking process in the [`wal`](https://github.com/ncruces/go-sqlite3/tree/wal) branch.

**Reply by ncruces** [reply]
The [wal](https://github.com/ncruces/go-sqlite3/tree/wal) branch has a _very_ preliminary version that ~_may_ work~ on macOS and Linux. Probably doesn't work at all (`mptest` is failing).

I need to go through [`os_unix.c`](https://github.com/sqlite/sqlite/blob/master/src/os_unix.c) and [`os_win.c`](https://github.com/sqlite/sqlite/blob/master/src/os_win.c) to understand the API better. It is very ill specified.

**Reply by infogulch** [reply]
Just to spot-check this I ran go-sqlite-bench with the current go.mod and with a [modified go.mod referencing the wal branch](https://github.com/infogulch/go-sqlite-bench/commit/a9c5621eb2b5b62dee16a9d477c1d7e2b9c26007). *I was not careful at all* but [there did not seem to be a step change in performance](https://gist.github.com/infogulch/bc1f8f488a6c9803b07f1311710e285a) between the versions where I might have expected the high concurrency benches to be noticeably better on wal. This could be caused by a number of things (an issue with my system, no bench designed to test wal perf, unnecessary locking, etc), but it didn't blow up at least. My test system is WSL2 ubuntu vm.

**Reply by ncruces** [reply]
That benchmark doesn't use WAL and only reads concurrently, so no change should be expected.

**Reply by infogulch** [reply]
I suspected that might be the case. What do you think would make a good wal benchmark? Maybe:

* Concurrently writing to N tables.
* In N threads: write a batch of 1k rows, read the whole table, repeat M times.

**Reply by ncruces** [reply]
SQLite doesn't do concurrent writes. Rollback journal mode works like a `RWMutex`: multiple readers, or a single writer, but not both. WAL allows multiple readers to coexist with a single writer.

**Reply by infogulch** [reply]
I'm thinking of this: https://sqlite.org/cgi/src/doc/begin-concurrent/doc/begin_concurrent.md

**Reply by ncruces** [reply]
That's not part of the main branch of SQLite and will not be supported, unless (maybe) you build SQLite yourself.

**Reply by infogulch** [reply]
Yes I was mistaken. I wrote a new benchmark to try to bench WAL mode:

https://github.com/infogulch/go-sqlite-bench/compare/master...bench-wal

```shell
./build-all.sh
./run-all.sh -simple=false -complex=false -many=false -large=false -concurrent=false -wal=true
```

It should read about 3x as much as it writes, and write across 1/4/16 goroutines.

It fails to run on several of the implementations, usually with a sqlite_busy error (the benchmark may need some tweaks). The benchmark for the wal branch of ncruces/go-sqlite3 appears to fail to read back rows after committing them, indicating corruption:

```
6_wal/1  - insert_query - ncruces    -      11124
6_wal/1  - dbsize       - ncruces    -   56573952
panic: must be equal but was different: 25000 != 1040

goroutine 99 [running]:
github.com/cvilsmeier/go-sqlite-bench/app.MustBeEqual(...)
        go-sqlite-bench/app/util.go:23
github.com/cvilsmeier/go-sqlite-bench/app.benchWal.func2.1(0xc001712000?)
        go-sqlite-bench/app/app.go:490 +0x568
github.com/cvilsmeier/go-sqlite-bench/app.benchWal.func2()
        go-sqlite-bench/app/app.go:505 +0x464
created by github.com/cvilsmeier/go-sqlite-bench/app.benchWal in goroutine 1
        go-sqlite-bench/app/app.go:475 +0x545
Sun Mar 31 14:49:18 CDT 2024
```

**Reply by ncruces** [reply]
The fact that `mptest` is failing, means the implementation is buggy. I'm not discarding the idea outright, but it's not a great sign. It may still be the case that I just messed up the locking, not the shared memory bits. Very few VFSes implement WAL, and it's very underspecified.

**Reply by ncruces** [reply]
Latest commit consistently passes `mptest`. This is good.
Must consider portability now.

**Reply by infogulch** [reply]
Great! With the latest commit, my benchmark doesn't panic or give invalid results anymore.

```
6_wal/1  - insert_query - ncruces    -      10245
6_wal/1  - dbsize       - ncruces    -   56573952
6_wal/2  - insert_query - ncruces    -       7493
6_wal/2  - dbsize       - ncruces    -   58912768
6_wal/4  - insert_query - ncruces    -       7574
6_wal/4  - dbsize       - ncruces    -   58814464
6_wal/8  - insert_query - ncruces    -       8066
6_wal/8  - dbsize       - ncruces    -   59703296
6_wal/16 - insert_query - ncruces    -       8544
6_wal/16 - dbsize       - ncruces    -   58777600
```

**Comment by ncruces** [comment]
This is being worked on https://github.com/ncruces/go-sqlite3/pull/71.
